### ISSUE TITLE
React Select Input Losing Focus under certain conditions

### ISSUE THREAD

**Issue by martinmckenna** [JedWatson/react-select#3575]
Repro case:

https://codesandbox.io/embed/nameless-leftpad-o3zwl

Basically what's happening in the code above is that I'm creating an array of react selects and adding and removing one when the add or remove buttons are clicked.

I'm using the custom input so that I can get a callback ref attached to the input DOM node and auto-focus it when the "add select" button is clicked.

The first issue

1. App loads
2. Cannot click into the first input

Second issue:

1. App loads
2. Click "add select"
3. Auto focus is applied to new select
4. Start typing and wait for the network request to resolve
5. Click "add select"
6. Click on previously typed-into select
7. Try typing
8. Observe - the input loses focus after the first character

The issue here is that when I create a new select, I cannot click on any of the previous ones before it, and sometimes when I type in the select that I _can_ type into, the select loses focus and focuses on the body after the network request resolves.

And here's my edited version without using the callback ref entirely:

https://codesandbox.io/embed/competent-cookies-tkfhc

The issue still remains

1. Try typing into input
2. After one character is typed into, the focus goes to the body

To be honest, I'm not sure if this is a bug with my code or the react-select callbacks, but I'm in need of some help.

**Comment by Rall3n** [comment]
tl;dr: The problem is your code.

The problem with your first issue is, that you have overwritten the `inputRef` prop for the `Input` component. `react-select` itself uses that prop to manage focus and blur. So removing that will let you focus the input on click again. An alternative to setting `inputRef` manually is using the `ref` prop on the `Select` itself. The `Select` component has a `focus` and `blur` method to manage exactly what you want to achieve.

The problem with you second issue is, that the `Input` component is re-created on each update, causing the user to lose focus on the input. Because of the solution for your first issue you don´t need to use a custom `Input` component, so the `components` prop can be left out. That should fix issue numero 2.

```js
const selectRef = React.useCallback(/* Your ref callback */);

<Select 
    ref={selectRef}
    {/* No `components` prop */}
    { ... }
/>
```
[Working example](https://codesandbox.io/s/reactselect3575-fixed-8u04m)

---

I would recommend to keep this sort of issue where you don´t know if it is your or libraries fault that the code is not working off GitHub and post that over at [Stack Overflow with the tag `react-select`](https://stackoverflow.com/questions/tagged/reactjs). If your question there resolves in an error of the library, you can open an issue here.

**Comment by martinmckenna** [comment]
yep that did the trick. Thanks!

**Comment by Slapbox** [comment]
What if you do need a custom `Input` component? 

This is a very aggravating and unexpected issue.

**Comment by tim-phillips** [comment]
@Slapbox, declare it outside the parent. If it relies on props, put it in a ref.

**Comment by brunofin** [comment]
this old ticket helped me solve an issue I was having, basically 1. I didn't need innerRef in my Input component, but I did need the custom Input component, so wrapping it around useMemo solved as well solved the issue.

**Comment by JoseVov** [comment]
> tl;dr: The problem is your code.
> 
> The problem with your first issue is, that you have overwritten the `inputRef` prop for the `Input` component. `react-select` itself uses that prop to manage focus and blur. So removing that will let you focus the input on click again. An alternative to setting `inputRef` manually is using the `ref` prop on the `Select` itself. The `Select` component has a `focus` and `blur` method to manage exactly what you want to achieve.
> 
> The problem with you second issue is, that the `Input` component is re-created on each update, causing the user to lose focus on the input. Because of the solution for your first issue you don´t need to use a custom `Input` component, so the `components` prop can be left out. That should fix issue numero 2.
> 
> ```js
> const selectRef = React.useCallback(/* Your ref callback */);
> 
> <Select 
>     ref={selectRef}
>     {/* No `components` prop */}
>     { ... }
> />
> ```
> 
> [Working example](https://codesandbox.io/s/reactselect3575-fixed-8u04m)
> 
> I would recommend to keep this sort of issue where you don´t know if it is your or libraries fault that the code is not working off GitHub and post that over at [Stack Overflow with the tag `react-select`](https://stackoverflow.com/questions/tagged/reactjs). If your question there resolves in an error of the library, you can open an issue here.

In my case, without overwritten the inputRef and just by adding onInputChange I can replicate this issue inversion 5.8.0. The Input Losing Focus after writing the first character.
